mathematica/retrobop_dynamical_stability.cpp

```diff
@@ -0,0 +1,668 @@
+}
+
+DegreesOfFreedom<Barycentric> EvaluateDegreesOfFreedom(
+    Ephemeris<Barycentric> const& ephemeris,
+    Celestial const celestial,
+    Instant const& t) {
+  return ephemeris.trajectory(ephemeris.bodies()[celestial])
+      ->EvaluateDegreesOfFreedom(t, /*hint=*/nullptr);
+}
+
+DegreesOfFreedom<Barycentric> JoolSystemBarycentre(
+    Ephemeris<Barycentric> const& ephemeris,
+    Instant const& t) {
+  BarycentreCalculator<DegreesOfFreedom<Barycentric>, GravitationalParameter>
+      jool_system_barycentre;
+  for (auto const celestial : jool_system) {
+    jool_system_barycentre.Add(
+        EvaluateDegreesOfFreedom(ephemeris, celestial, t),
+        ephemeris.bodies()[celestial]->gravitational_parameter());
+  }
+  return jool_system_barycentre.Get();
+}
+
+not_null<std::unique_ptr<Ephemeris<Barycentric>>> MakeEphemeris(
+    HierarchicalSystem<Barycentric>::BarycentricSystem&& system,  // NOLINT
+    FixedStepSizeIntegrator<
+        Ephemeris<Barycentric>::NewtonianMotionEquation> const& integrator,
+    Time const& step) {
+  return std::make_unique<Ephemeris<Barycentric>>(
+      std::move(system.bodies),
+      system.degrees_of_freedom,
+      ksp_epoch,
+      /*fitting_tolerance=*/1 * Milli(Metre),
+      Ephemeris<Barycentric>::FixedStepParameters(integrator, step));
+}
+
+template<typename BitGenerator>
+Vector<double, Barycentric> RandomUnitVector(BitGenerator& generator) {
+  std::uniform_real_distribution<> longitude_distribution(-π, π);
+  std::uniform_real_distribution<> z_distribution(-1, 1);
+  double const z = z_distribution(generator);
+  Angle const longitude = longitude_distribution(generator) * Radian;
+  return Vector<double, Barycentric>({Cos(longitude) * Sqrt(1 - Pow<2>(z)),
+                                      Sin(longitude) * Sqrt(1 - Pow<2>(z)),
+                                      z});
+}
+
+template<typename Integrator>
+std::list<not_null<std::unique_ptr<Ephemeris<Barycentric>>>>
+MakePerturbedEphemerides(int const count,
+                         Integrator const& integrator,
+                         Time const& step) {
+  std::mt19937_64 generator;
+  std::list<not_null<std::unique_ptr<Ephemeris<Barycentric>>>> result;
+  for (int i = 0; i < count; ++i) {
+    auto system = ReadSystem(system_file);
+    for (auto const celestial : jool_system) {
+      system.degrees_of_freedom[celestial] = {
+          system.degrees_of_freedom[celestial].position() +
+              RandomUnitVector(generator) * Milli(Metre),
+          system.degrees_of_freedom[celestial].velocity()};
+    }
+    result.emplace_back(MakeEphemeris(std::move(system), integrator, step));
+  }
+  return result;
+}
+
+void FillPositions(
+    Ephemeris<Barycentric> const& ephemeris,
+    Instant const& initial_time,
+    Time const& duration,
+    std::vector<std::vector<Vector<double, Barycentric>>>& container) {
+  for (int n = 0; n * step < duration; ++n) {
+    Instant const t = initial_time + n * step;
```

Comment:
Make `n * step` the loop variable, not `n`.  We are not writing FORTRAN this week.